### PR TITLE
Attachment Downloader

### LINKED ISSUE #1040: Allow download of images shared in chat
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/master/CONTRIBUTING.md) for onboarding!
___

**Platform - version:**  
All platforms / 280

**Action Performed (reproducible steps):**
1. Log into Expensify Cash
2. Click the paperclip icon
3. Select an image and confirm the upload
4. Once the image uploads to the chat, click the image
5. Notice that there is no way to download the image in the Attachment modal

**Expected Result:**
The modal should provide an icon that can be clicked to download the image.

**Actual Result:**
There is no way to download the image

**Notes/Photos/Videos:**
 
 Here is what it currently looks like:
image

Instead, we'd like to include a download button:
101112969-47713c00-3593-11eb-89ed-02b61f92bcbc-1

The icon asset can be found here:
101107234-ba27ea80-3586-11eb-940c-024ae39438ca

After downloading, please confirm the image is stored on the device.

view the job on Upwork [here](https://www.upwork.com/jobs/~01340639ab0b967185).

### PR DESCRIPTION
**Pull request by daniyal-intellicel**
@jboniface, @laurenreidexpensify  please review this PR for issue [1040](https://github.com/Expensify/Expensify.cash/issues/1040)

### Details
<Explanation of the change or anything fishy that is going on>
The attachment modal in chat now provides an icon that can be clicked to download the image opened in modal.

### Fixed Issues
<Please replace GH_LINK with the link to the GitHub issue this Pull Request is fixing>
Fixes /issues/1040

### Tests

The attachment download feature was tested on below mentioned platforms and following steps can be followed to perform the test on any platform. This feature could not be tested on physical iOS device because its testing require provisioning profile that should include uuid of the testing device. So, any contributor who has that profile can help testing it.
1. Log into Expensify Cash
2. Click the paperclip icon
3. Select an image and confirm the upload
4. Once the image uploads to the chat, click the image to open in modal
5. Click on the download icon at the top right corner of the modal
6. Verify that image gets downloaded 
    a. For mobile apps check downloads/documents folder of device
    b. For desktop app the user will be prompted to save the file in desired folder
    c. For web app check the download manager of the browser

### Tested On

- [x] Web
- [ ] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

### Screenshots
<Add any necessary screenshots for all platforms if your change added or updated UI>

attachment-downloader-web-chrome

#### Web

(Chrome)
attachment-downloader-web-chrome

#### Desktop
(MacOS)
attachment-downloader-desktop-mac

#### iOS

(Simulator)
attachment-downloader-iOS-simulator

#### Android

(Physical Device)
attachment-downloader-android-device

**I have read the CLA Document and I hereby sign the CLA**

### CODE REVIEW

**Comment by daniyal-intellicel** [comment]
**I have read the CLA Document and I hereby sign the CLA**

**Comment by daniyal-intellicel** [comment]
**recheck**

File: ios/Podfile.lock
Comment (deetergp) [inline ios/Podfile.lock:675]: Did you mean to roll back Cocoapods back?

File: src/libs/downloadUrlFile/index.native.js
Comment (deetergp) [inline src/libs/downloadUrlFile/index.native.js:37]: Does `console.log()` even work in native? Regardless, if you're not doing anything inside `.then()` or `.catch()` then let's get rid of them.

**Review by deetergp — CHANGES_REQUESTED** [review]
Confirmed working in Web & Desktop, but neither iOS nor Android work. When clicking the download icon in iOS nothing appears to happen, I'm not prompted for anything, and the image never appears in Photos.app. In Android, when I click the icon, I get the following errors message:
Screen Shot 2021-01-12 at 4 40 49 PM

Don't both Android and iOS require permissions from their respective settings in order for apps to access their photo storage?

File: src/libs/downloadUrlFile/index.native.js
Comment (daniyal-intellicel) [reply src/libs/downloadUrlFile/index.native.js:37]: Yes, `console.log()` does work in native. I added `.then()` and `.catch()` for testing purpose and you are right that they are not doing anything so i'll remove them.

File: ios/Podfile.lock
Comment (daniyal-intellicel) [reply ios/Podfile.lock:675]: I didn't change the lock file. It was an automatic change that was generated by installing pods in ios project.

File: src/libs/downloadUrlFile/index.native.js
Comment (deetergp) [reply src/libs/downloadUrlFile/index.native.js:37]: 🤦 Well that's embarrassing, of course native has a JS console. Thanks for the reminder.

**Comment by deetergp** [comment]
Am I missing something with regard to testing the native/mobile versions of this? P
If I am unable to see that it works on all platforms, I can't merge it. Additionally, you've got a merge conflict that needs to be resolved. Thanks!

**Comment by daniyal-intellicel** [comment]
> Am I missing something with regard to testing the native/mobile versions of this? P
> If I am unable to see that it works on all platforms, I can't merge it. Additionally, you've got a merge conflict that needs to be resolved. Thanks!

I guess native permission for storage was not granted on your dev devices. I'll add permission request dialogue in download handler and merge conflicts also.

**Comment by deetergp** [comment]
> I guess native permission for storage was not granted on your dev devices. I'll add permission request dialogue in download handler and merge conflicts also.
That would be great. Thanks!

**Comment by daniyal-intellicel** [comment]
@deetergp I have added request dialogue to ask user for storage permission. But I'm unable to fix conflicting files. How can I fix a conflict when I don't have write access to the repo.

**Comment by deetergp** [comment]
@daniyal-intellicel Why do you need write access to the repo? You should pull `master` from the base repo into your local master, merge your local master into your feature branch, then resolve the conflicts there.

**Comment by deetergp** [comment]
When testing iOS, after re-building your branch, I was prompted to grant permission to my photos the first time I clicked the download button, but when I checked my Photos app, the image was not there. Subsequent clicks of the download button did not prompt me for permission, which is expected, but still produced no downloaded image.

Can you please detail the exact steps you took to test this locally in your native iOS and Android apps?

### VidCap
[iOS Download Attempt](
https://user-images.githubusercontent.com/766197/106948467-99902400-66e0-11eb-9ff4-af6984cecb39.mp4)

### Permissions Prompt
Screen Shot 2021-02-04 at 11 51 15 AM

**Comment by deetergp** [comment]
Any updates, @daniyal-intellicel?

**Comment by daniyal-intellicel** [comment]
> Any updates, @daniyal-intellicel?

Yeah, I'm looking for the problem

**Comment by daniyal-intellicel** [comment]
> Any updates, @daniyal-intellicel?

Yeah, I'm looking for the problem

**Comment by daniyal-intellicel** [comment]
> Any updates, @daniyal-intellicel?

Yeah, I'm looking for the problem

**Pull request closed without merging**
